fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,822 @@
+                        result = self.callbacks(self, self.step_)
+                        if result and not isinstance(result, bool):
+                            raise RuntimeError(_CALLBACK_RETURNS_ERROR)
+                        stop = result
+
+                    if stop:
+                        return self
+            if self.shuffle and epoch != epochs - 1:  # Don't shuffle last epoch
+                X, Y, Z = self.backendEngine_.shuffle(X, Y, Z)
+
+        return self
+
+    def partial_fit(self, X, y, *, sensitive_features=None):
+        """
+        Perform one epoch on given samples and update model.
+
+        Parameters
+        ----------
+        X : numpy.ndarray
+            Two-dimensional numpy array containing training data
+
+        y : array
+            Array-like containing training targets
+
+        sensitive_features : array
+            Array-like containing the sensitive feature of the
+            training data.
+        """
+        X, Y, Z = self._validate_input(
+            X, y, sensitive_features, reinitialize=False
+        )
+        self.backendEngine_.train_step(X, Y, Z)
+
+        return self
+
+    def decision_function(self, X):
+        """
+        Compute predictor output for given test data.
+
+        Parameters
+        ----------
+        X : numpy.ndarray
+            Two-dimensional numpy array containing test data
+
+        Returns
+        -------
+        Y_pred : numpy.ndarray
+            Two-dimensional array containing the model's (soft-)predictions
+        """
+        check_is_fitted(self)
+        X = check_X(X)
+
+        Y_pred = self.backendEngine_.evaluate(X)
+        return Y_pred
+
+    def predict(self, X):
+        """
+        Compute discrete predictions for given test data.
+
+        Parameters
+        ----------
+        X : numpy.ndarray
+            Two-dimensional numpy array containing test data
+
+        Returns
+        -------
+        Y_pred : array
+            array-like containing the model's predictions fed through
+            the (discrete) :code:`predictor_function`
+        """
+        check_is_fitted(self)
+        X = check_X(X)
+
+        Y_pred = self.backendEngine_.evaluate(X)
+        Y_pred = self.predictor_function_(Y_pred)
+        Y_pred = self.y_transform_.inverse_transform(Y_pred)
+        return Y_pred
+
+    def _validate_input(self, X, Y, Z, reinitialize=False):
+        """
+        Validate the input data and possibly setup this estimator.
+
+        Important note is that we follow call `__setup` from here, because the
+        setup procedure requires the validated data. If `reintialize` is True,
+        then always call `__setup`.
+        """
+        if not self.skip_validation:
+            X = check_X(X)
+
+        try:  # TODO check this
+            check_is_fitted(self)
+            is_fitted = True
+        except NotFittedError:
+            is_fitted = False
+
+        if Z is None:
```

Comment:
Don't we use `A` for sensitive features?

Reply:
Yes, I mistakenly used the paper's conventions